File: 1514/solution.ts
function maxProbability(
  n: number,
  edges: number[][],
  succProb: number[],
  start_node: number,
  end_node: number
): number {
  const maxQ = new MaxPriorityQueue({
    compare: (a, b) => b[0] - a[0],
  });

  const paths: Map<number, [number, number][]> = new Map();
  for (let i = 0; i < edges.length; i++) {
    const [start, end] = edges[i];
    if (!paths.get(start)) {
      paths.set(start, []);
    }
    if (!paths.get(end)) {
      paths.set(end, []);
    }
    paths.get(start)!.push([succProb[i], end]);
    paths.get(end)!.push([succProb[i], start]);
  }
  const visited: Set<number> = new Set();
  maxQ.enqueue([1, start_node]);
  while (maxQ.size()) {
    const [dist, node] = maxQ.dequeue();
    console.log(dist, node);
    if (visited.has(node)) {
      continue;
    }

    if (node === end_node) {
      return dist;
    }
    visited.add(node);
    if (!paths.get(node)) {
      continue;
    }

    for (const [dist2, child] of paths.get(node)!) {
      if (!visited.has(child)) {
        maxQ.enqueue([dist * dist2, child]);
      }
    }
  }

  return 0;
}

File: 1514/solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

class Solution {
    public double maxProbability(int n, int[][] edges, double[] succProb, int start_node, int end_node) {
        PriorityQueue<Tuple> heap = new PriorityQueue<>(new Tuple.TupleComparator());
        boolean[] visited = new boolean[n];
        Arrays.fill(visited, false);
        Map<Integer, List<Tuple>> graph = new HashMap<>();
        for (int i = 0; i < edges.length; i++) {
            int[] edge = edges[i];
            int x = edge[0];
            int y = edge[1];
            graph.computeIfAbsent(x, k -> new ArrayList<>()).add(new Tuple(succProb[i], y));
            graph.computeIfAbsent(y, k -> new ArrayList<>()).add(new Tuple(succProb[i], x));
        }
        heap.add(new Tuple(1, start_node));
        while (!heap.isEmpty()) {
            Tuple curr = heap.poll();
            if (curr.node == end_node) {
                return curr.prob;
            }
            if (visited[curr.node]) {
                continue;
            }
            visited[curr.node] = true;
            if (graph.containsKey(curr.node)) {
                for (Tuple child : graph.get(curr.node)) {
                    if (!visited[child.node]) {
                        heap.add(new Tuple(child.prob * curr.prob, child.node));
                    }
                }
            }
        }
        return 0;
    }
}

class Tuple {
    double prob;
    int node;

    public Tuple(double prob, int node) {
        this.prob = prob;
        this.node = node;
    }

    public static class TupleComparator implements Comparator<Tuple> {
        @Override
        public int compare(Tuple t1, Tuple t2) {
            return Double.compare(t2.prob, t1.prob);
        }
    }
}